File: go/1600.go
package main

type ThroneInheritance struct {
	king  string
	edges map[string][]string
	dead  map[string]struct{}
}

func Constructor(kingName string) ThroneInheritance {
	return ThroneInheritance{kingName, map[string][]string{}, map[string]struct{}{}}
}

func (t *ThroneInheritance) Birth(parentName string, childName string) {
	t.edges[parentName] = append(t.edges[parentName], childName)
}

func (t *ThroneInheritance) Death(name string) {
	t.dead[name] = struct{}{}
}

func (t *ThroneInheritance) GetInheritanceOrder() []string {
	ans := []string{}
	var preOrder func(name string)
	preOrder = func(name string) {
		if _, ok := t.dead[name]; !ok {
			ans = append(ans, name)
		}
		for _, child := range t.edges[name] {
			preOrder(child)
		}
	}
	preOrder(t.king)
	return ans
}

/**
 * Your ThroneInheritance object will be instantiated and called as such:
 * obj := Constructor(kingName);
 * obj.Birth(parentName,childName);
 * obj.Death(name);
 * param_3 := obj.GetInheritanceOrder();
 */

File: python/1600.py
from collections import defaultdict
from typing import List

class ThroneInheritance:

    def __init__(self, kingName: str):
        self.childs = defaultdict(list)
        self.dead = set()
        self.king = kingName

    def birth(self, parentName: str, childName: str) -> None:
        self.childs[parentName].append(childName)

    def death(self, name: str) -> None:
        self.dead.add(name)

    def getInheritanceOrder(self) -> List[str]:
        ans = list()
        def perOrder(name: str) -> None:
            if name not in self.dead:
                ans.append(name)
            for child in self.childs[name]:
                perOrder(child)
        perOrder(self.king)
        return ans

# Your ThroneInheritance object will be instantiated and called as such:
# obj = ThroneInheritance(kingName)
# obj.birth(parentName,childName)
# obj.death(name)
# param_3 = obj.getInheritanceOrder()